Parse/src/main/java/com/parse/ParseObjectSubclassingController.java

```diff
@@ -0,0 +1,133 @@
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+package com.parse;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Modifier;
+import java.util.HashMap;
+import java.util.Map;
+
+/* package */ class ParseObjectSubclassingController {
+  private static class ObjectSubclassInfo {
+    public final Class<? extends ParseObject> type;
+    public final String className;
+    public final Constructor<? extends ParseObject> constructor;
+
+    public ObjectSubclassInfo(Class<? extends ParseObject> clazz) throws NoSuchMethodException {
+      type = clazz;
+      className = getClassName(type);
+      constructor = getConstructor(clazz);
+    }
+
+    private static Constructor<? extends ParseObject> getConstructor(Class<? extends  ParseObject> clazz) throws NoSuchMethodException {
+      Constructor<? extends ParseObject> constructor = clazz.getDeclaredConstructor();
+      int modifiers = constructor.getModifiers();
+      if (Modifier.isPublic(modifiers) || (clazz.getPackage().getName().equals("com.parse") &&
+          !(Modifier.isProtected(modifiers) || Modifier.isPrivate(modifiers)))) {
+        return constructor;
+      }
+      throw new NoSuchMethodException();
+    }
+
+    private static String getClassName(Class<? extends ParseObject> clazz) {
+      ParseClassName info = clazz.getAnnotation(ParseClassName.class);
+      if (info == null) {
+        return null;
+      }
+      return  info.value();
+    }
+  }
+
+  private Object mutex;
+  private Map<String, ObjectSubclassInfo> registeredSubclasses;
+
+  public ParseObjectSubclassingController() {
+    mutex = new Object();
+    registeredSubclasses = new HashMap<>();
+  }
+
+  String getClassName(Class<? extends ParseObject> clazz) {
+    return ObjectSubclassInfo.getClassName(clazz);
+  }
+
+  boolean isTypeValid(String className, Class<? extends ParseObject> clazz) {
+    ObjectSubclassInfo info = null;
+
+    synchronized (mutex) {
+      info = registeredSubclasses.get(className);
+    }
+
+    return info == null
+      ? clazz == ParseObject.class
+      : info.type == clazz;
+  }
+
+  void registerSubclass(Class<? extends ParseObject> clazz) {
+    if (!ParseObject.class.isAssignableFrom(clazz)) {
+      throw new IllegalArgumentException("Cannot register a type that is not a subclass of ParseObject");
+    }
+
+    String className = ObjectSubclassInfo.getClassName(clazz);
+
+    synchronized (mutex) {
+      ObjectSubclassInfo previousInfo = registeredSubclasses.get(className);
+      if (previousInfo != null) {
+        if (clazz.isAssignableFrom(previousInfo.type)) {
+          // Previous subclass is more specific or equal to the current type, do nothing.
+          return;
+        } else if (previousInfo.type.isAssignableFrom(clazz)) {
+          // Previous subclass is parent of new child subclass, fallthrough and actually
+          // register this class.
+                    /* Do nothing */
```

Comment:
This is intentional fallthrough, when the old registration is a parent of the new registration:

Hierarchy: ParseObject -> A ->  B -> C

If we have 'A' registered, and now we register 'B', we should fall through inthis case and perform the operation.